### PR TITLE
apply most of python futurist stage1 tools for python3 compatibility

### PR DESCRIPTION
**Pull request by davidlange6**
-f lib2to3.fixes.fix_apply
-f lib2to3.fixes.fix_except
-f lib2to3.fixes.fix_exitfunc
-f lib2to3.fixes.fix_funcattrs
-f lib2to3.fixes.fix_has_key
-f lib2to3.fixes.fix_idioms
-f lib2to3.fixes.fix_intern
-f lib2to3.fixes.fix_isinstance
-f lib2to3.fixes.fix_methodattrs
-f lib2to3.fixes.fix_ne
-f lib2to3.fixes.fix_numliterals
-f lib2to3.fixes.fix_paren
-f lib2to3.fixes.fix_reduce
-f lib2to3.fixes.fix_renames
-f lib2to3.fixes.fix_repr
-f lib2to3.fixes.fix_standarderror
-f lib2to3.fixes.fix_sys_exc
-f lib2to3.fixes.fix_throw
-f lib2to3.fixes.fix_tuple_params
-f lib2to3.fixes.fix_types
-f lib2to3.fixes.fix_xreadlines
-f libfuturize.fixes.fix_absolute_import
-f libfuturize.fixes.fix_next_call
-f libfuturize.fixes.fix_raise

["most" in terms of tools, but the remaining ones affect more code]

Other than the changes to argparse.py in RecoLuminosity changes are minor - the argparse file looks to be obsolete in any case. I'll look at removing it in favor of the standard one...

### CODE REVIEW

File: CalibTracker/SiStripDCS/test/ParseManualO2Olog.py [hunk 1]
```diff
@@ -731,9 +731,8 @@ def getIOVsInTimeInterval(self,StartTime,StopTime,HistoryDict):
         Function that returns the IOV timestamps (ordered) contained in a given interval.
         """
         #Copy the timestamps in a list
-        TimeStamps=HistoryDict.keys()[:]
+        TimeStamps=sorted(HistoryDict.keys()[:])
         #Sort them:
```
Comment (fwyzard) [inline CalibTracker/SiStripDCS/test/ParseManualO2Olog.py]: the comment should be moved to the previous line ?

File: CalibTracker/SiStripDCS/test/ParseManualO2Olog.py [hunk 2]
```diff
@@ -750,9 +749,8 @@ def getReducedIOVs (self,StartTime,StopTime,HistoryDict,deltaT,maxIOVLength):
         deltaTime=datetime.timedelta(seconds=deltaT)
         maxSequenceLength=datetime.timedelta(seconds=maxIOVLength)
         #Copy the timestamps in a list:
-        TimeStamps=HistoryDict.keys()[:]
+        TimeStamps=sorted(HistoryDict.keys()[:])
         #Sort them:
```
Comment (fwyzard) [inline CalibTracker/SiStripDCS/test/ParseManualO2Olog.py]: ditto

File: DQM/Integration/scripts/harvesting_tools/cmsHarvester.py [hunk 3]
```diff
@@ -3409,10 +3406,9 @@ def build_dataset_list(self, input_method, input_name):
         # NOTE: There should not be any duplicates in any list coming
         # from DBS, but maybe the user provided a list file with less
         # care.
-        dataset_names = list(set(dataset_names))
+        dataset_names = sorted(set(dataset_names))
 
         # Store for later use.
```
Comment (fwyzard) [inline DQM/Integration/scripts/harvesting_tools/cmsHarvester.py]: ditto

File: DQMServices/FileIO/scripts/dqmMemoryStats.py [hunk 4]
```diff
@@ -90,10 +90,10 @@ def displayDirectoryStatistics(stats, args):
     cutoff, display = args.cutoff * 1024, args.display
 
     as_list = [(v, k, ) for (k, v) in group_stats.items()]
-    as_list.sort(reverse=True, key=lambda (v, k): abs(v))
+    as_list.sort(reverse=True, key=lambda v_k1: abs(v_k1[0]))
```
Comment (fwyzard) [inline DQMServices/FileIO/scripts/dqmMemoryStats.py]: I think the old syntax is clearer here. Is there a problem with it ?

File: DQMServices/FileIO/scripts/dqmMemoryStats.py [hunk 5]
```diff
@@ -90,10 +90,10 @@ def displayDirectoryStatistics(stats, args):
...
 
     if cutoff is not None:
-        as_list = filter(lambda (v, k): abs(v) > cutoff, as_list)
+        as_list = filter(lambda v_k: abs(v_k[0]) > cutoff, as_list)
```
Comment (fwyzard) [inline DQMServices/FileIO/scripts/dqmMemoryStats.py]: ditto

**Comment by kpedro88** [comment]
+1

File: PhysicsTools/PythonAnalysis/python/rootplot/argparse.py [hunk 6]
```diff
@@ -1,2353 +1,2352 @@
-# -*- coding: utf-8 -*-
```
Comment (fwyzard) [inline PhysicsTools/PythonAnalysis/python/rootplot/argparse.py]: Shouldn't we be able to drop this file entirely by now ?

File: RecoLuminosity/LumiDB/python/argparse.py [hunk 7]
```diff
@@ -1,2270 +1,2269 @@
-# -*- coding: utf-8 -*-
```
Comment (fwyzard) [inline RecoLuminosity/LumiDB/python/argparse.py]: ditto

**Comment by dmitrijus** [comment]
+1

File: DQMServices/FileIO/scripts/dqmMemoryStats.py (hunk 4 above)
Comment (davidlange6) [reply DQMServices/FileIO/scripts/dqmMemoryStats.py]: I don't know. Will look into it.

Comment (dmitrijus) [reply DQMServices/FileIO/scripts/dqmMemoryStats.py]: It is indeed clearer. I don't know why it does this replacement.

File: PhysicsTools/PythonAnalysis/python/rootplot/argparse.py (hunk 6 above)
Comment (davidlange6) [reply PhysicsTools/PythonAnalysis/python/rootplot/argparse.py]: I commented on that in my PR (well, for the RecoLuminosity one)..

File: DQMServices/FileIO/scripts/dqmMemoryStats.py (hunk 4 above)
Comment (davidlange6) [reply DQMServices/FileIO/scripts/dqmMemoryStats.py]: here is the reason
https://www.python.org/dev/peps/pep-3113/

(I found this a bit more explanatory: https://stackoverflow.com/questions/15712210/python-3-2-lambda-syntax-error). 

and indeed the old syntax throws and error in python3

Python 3.6.4 (default, Apr 27 2018, 10:48:00) 
[GCC 6.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> dud=[(1,5),(2,3),(3,2),(4,20),(5,-1)]
>>> dud.sort(reverse=True,key=lambda (v,k): abs(v))
  File "<stdin>", line 1
    dud.sort(reverse=True,key=lambda (v,k): abs(v))
                                     ^
SyntaxError: invalid syntax

**Comment by davidlange6** [comment]
please test

**Comment by Dr15Jones** [comment]
+1

**Comment by kpedro88** [comment]
Same failure as in #23205

**Comment by kpedro88** [comment]
reco changes in miniAOD output... odd.

**Comment by perrotta** [comment]
CMSSW_10_2_X_2018-05-15-1100 was still used as baseline: another run of tests will be needed if we want to get rid of the spurious extra commits

**Comment by perrotta** [comment]
please test

**Comment by davidlange6** [comment]
> On May 17, 2018, at 12:39 PM, cmsbuild <notifications@github.com> wrote:
> 
> Comparison is ready
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23208/28000/summary.html
> 
> Comparison Summary:
> 
> 	• No significant changes to the logs found
> 	• Reco comparison results: 6 differences found in the comparisons
> 	• DQMHistoTests: Total files compared: 30
> 	• DQMHistoTests: Total histograms compared: 2740553
> 	• DQMHistoTests: Total failures: 2218

I was hunting for these differences - there are indeed a bunch of histograms identified as different in the comparisons - but I am yet to see a plot that is actually different. Is there an unrelated problem?

There are JR test differences in the 2011 HI test, but I see that issue also in #23160?

**Comment by makortel** [comment]
@davidlange6 Are the differences perhaps those discussed in #23105?

**Comment by davidlange6** [comment]
looks like it -

**Comment by Dr15Jones** [comment]
@davidlange6 The `testmakepset::fileinpathTest` test failure is due to a test which is trying to read a file it only thinks is in the base release area and not in the working directory. That file is `Geometry/TrackerSimData/data/trackersens.xml` (I assumed picked for some random reason). Most of the time it is true that code which requires `PythonParameterSet` to be recompiled will not also require checking out `Geometry/TrackerSimData`. However, for your particular case, that assumption did not hold.

**Comment by davidlange6** [comment]
nice - so I think this PR is ready then. Failing tests are false alarms.

**Comment by kpedro88** [comment]
+1

**Comment by perrotta** [comment]
+1
- Technical
- In the reco related part it uses isinstance() and sorted() python methods, and import argparse from default python library
- Jenkins tests pass, no difference observed

**Comment by dmitrijus** [comment]
+1

**Comment by Martin-Grunewald** [comment]
+1

**Comment by fabozzi** [comment]
+1

**Comment by rekovic** [comment]
+1

**Comment by ggovi** [comment]
+1

**Comment by arunhep** [comment]
+1

**Comment by fabiocos** [comment]
@ssekmen @civanch could you please sign so as we move forward? Or do you see problems?

**Comment by ssekmen** [comment]
+1

**Comment by fabiocos** [comment]
+operations

the changes to the StandardSequences look transparent, according to the syntax update

**Comment by fabiocos** [comment]
@davidlange6 @smuzaffar not sure who is Mr. cmsdoxy who should sign for docs, anyway the change looks transparent according to the syntax update, so I will just merge as soon as @civanch gives his green light

**Comment by civanch** [comment]
+1

**Comment by fabiocos** [comment]
@davidlange6 you are correct, 8 days ago, I take it for good

**Comment by fabiocos** [comment]
@civanch thanks

**Comment by fabiocos** [comment]
please test

**Comment by cmsdoxy** [comment]
+1

**Comment by fabiocos** [comment]
+1

residual differences in comparisons look unrelated to this PR but are the same observed in  the overlapping #23065

**Comment by fabiocos** [comment]
merge

**Pull request merged**